### DISCUSSION TITLE
BlackJAX & Probabilistic Numerical ODE Solvers

### DISCUSSION THREAD

**Discussion by pnkraemer** [blackjax-devs/blackjax#497] [Show and tell]
Hey guys,

Big fan of BlackJAX here! I recently published some [probabilistic numerics JAX code](https://pnkraemer.github.io/probdiffeq/). I made a [tutorial notebook](https://pnkraemer.github.io/probdiffeq/advanced_examples/physics_enhanced_regression_2/) about combining BlackJAX with probabilistic ODE solvers. The sampling bit worked like a charm!

Wanted to share this in case probabilistic numerics or differential equations are of interest to anyone here :)

Nico

**Comment by albcab** [comment]
This is amazing! And a unique use of the library. It would be great to have your notebook featured on the [Sampling book](https://github.com/blackjax-devs/sampling-book) if possible. Notice `window_adaptation` returns parameters of the kernel and not the kernel itself on the latest version of the library :smile:

**Reply by rlouf** [reply]
Awesome, this kind of application is exactly what I had in mind when designing Blackjax. Happy to hear it's helpful :)

**Reply by pnkraemer** [reply]
Awesome, glad you like it! :)  (And thanks for the `window_adaptation` pointer.)


> It would be great to have your notebook featured on the [Sampling book](https://github.com/blackjax-devs/sampling-book) if possible.

Sure. What would we need to do?

**Reply by albcab** [reply]
Normally you would open an issue on blackjax-devs/sampling-book but since you already have a working markdown file, skip this and open a PR directly.

Place the `.md` file in the `sampling-book/book/models` directory, maybe change its title to highlight its use for probabilistic numerical ODE solvers. It would be great if you could include some of the details in part 1 of your implementation for better context, like the bibliography (follow the example of [Periodic Orbital MCMC](https://github.com/blackjax-devs/sampling-book/blob/main/book/algorithms/PeriodicOrbitalMCMC.md)). Call the file from the [TOC](https://github.com/blackjax-devs/sampling-book/blob/main/book/_toc.yml). Then build the docs on you local machine to make sure it all runs correctly:
```python
pip install -r requirements.txt
jupyter-book build --all book/
```
(you might need to `mkdir jupyter_cache` if there isn't one), and open a PR. Thanks!

**Reply by pnkraemer** [reply]
PR has been opened. Thanks for the explanations, @albcab, that accelerated the entire process massively :)

**Comment by junpenglao** [comment]
🤖 *This reply was drafted with Claude and reviewed by me (@junpenglao).*

This is a great use of the library — we'd still love to feature it in the [sampling-book](https://github.com/blackjax-devs/sampling-book). One API note since this thread is a bit old: `window_adaptation.run` now returns `(state, parameters), info` (parameters, not a kernel), and you can drive the sampling loop with `blackjax.util.run_inference_algorithm`.

**Reply by pnkraemer** [reply]
Thanks for messaging! 

Not sure I entirely understand: are you saying the current code (respectively [Probdiffeq's version](https://pnkraemer.github.io/probdiffeq/examples_advanced/parameter_estimation_blackjax/)) breaks now, will break in the next release, or works fine but could be optimised? :)

**Reply by junpenglao** [reply]
Disregard this message - the ODE example works fine with probdiffeq in sampling-book (https://blackjax-devs.github.io/sampling-book/models/probabilistic-ode-solver-parameter-estimation/). The LLM replies was instructed by a prompt to check discussion and reply if the old posts provide information outdated due to recent API changes. I am going to ask Claude why they missed the context here.

**Reply by junpenglao** [reply]
LOL the reason was that we just read body + top-level comment only and never traversed the nested reply chain - lesson learned.

@pnkraemer how have you been? any exciting update on probdiffeq or numerical ode land in general?

**Reply by pnkraemer** [reply]
Oh, ok -  glad everything works then!

That said, there will be some changes to the probabilistic numerics code in the next release (to enable some new features, like solving different kinds of ODEs). I left a reminder on Probdiffeq to revisit the sampling book example here once it's released (https://github.com/pnkraemer/probdiffeq/issues/876).
